### DISCUSSION TITLE
`is var` and `is not var` work with if but not while loop

### DISCUSSION THREAD

**Discussion by Thaina** [dotnet/csharplang#7676] [Language Ideas]
```C#
using System;
using System.Threading;
					
public class Program
{
	public static void Main()
	{
		while(ShouldWork() is not bool apply)
		{
			Thread.Sleep(0);
		}
		
		Console.WriteLine(apply);
	}
	
	static bool condition = true;
	static bool? ShouldWork()
	{
		if(!condition)
			return null;
		
		return true;
	}
}
```

This code above give error

> Compilation error (line 13, col 21): The name 'apply' does not exist in the current context

https://dotnetfiddle.net/vi1hxZ

As the logic are very simple and obvious branch prediction, `while` condition without internal `break` should allow initialize variable like `if`

Is it possible?

**Comment by HaloFour** [comment]
Pattern variables declared within a `while` condition are scoped to the loop body only.

**Reply by TahirAhmadov** [reply]
I think I may have read something about it, but can't remember the details now. What was the justification for it, do you know?

**Reply by HaloFour** [reply]
See: https://github.com/dotnet/csharplang/blob/main/meetings/2016/LDM-2016-11-30.md#scope-of-while-condition-expression-variables

and: https://github.com/dotnet/roslyn/issues/15529

**Reply by Thaina** [reply]
I see. Thank you for your explanation. But still `is not` was the opposite of the case though. It will never use in the loop and only available outside of the loop

**Reply by HaloFour** [reply]
The pattern doesn't change the scoping of the pattern variable.  It is always scoped to the body of the loop.  C# doesn't have any concept of "flow scoping" where the scope of the variable follows definite assignment.
